### ISSUE TITLE
Call for Layer Layer Normalization

### ISSUE THREAD

**Issue by icemansina** [Lasagne/Lasagne#736]
I am trying to implement the layer normalization and LNRNN. Is there anybody interested in this implementation in Lasagne?

PS: Keras code: https://gist.github.com/udibr/7f46e790c9e342d75dcbd9b1deb9d940
      Tensorflow code: https://github.com/pbhatia243/tf-layer-norm/blob/master/layers.py

@f0k

**Comment by erfannoury** [comment]
I have a working implementation of Layer Normalization for LSTM [(LN-LSTM)](https://github.com/erfannoury/seq2seq-lasagne/blob/master/CustomLSTMLayer.py#L17-L650) that you may take a look. We can further improve it if you notice anything that needs to be changed.

**Comment by f0k** [comment]
> Is there anybody interested in this implementation in Lasagne?

The existing `BatchNormLayer` can already do layer normalization by just changing the `axis` specification, but then you'd always need to call `get_output()` with `batch_norm_use_averages=False, batch_norm_compute_averages=False`. So we could either add a `LayerNormLayer` (uh...) or make the running averages of `BatchNormLayer` completely optional and just add a `layernorm()` convenience function that instantiates a `BatchNormLayer`. Hmm, or maybe add a `NormLayer` for layer normalization and derive `BatchNormLayer` from it, just adding the running average statistics on top and changing the default `axis`. I think that might be the cleanest option. Sorry for thinking aloud.

For recurrent models, currently the only solution are separate classes like an LNLSTMLayer, and I'm not sure we want to add this to Lasagne, as it duplicates a lot of code. We'll be able to include a much more modular solution with #629, but this is blocked by #678, which is blocked by https://github.com/Lasagne/Lasagne/milestone/2 -- any help with this would be appreciated!

**Comment by icemansina** [comment]
Thanks a lot. @f0k @erfannoury 

- For normal model, i.e., CNN or Dense layers, 

> So we could either add a `<LayerNormLayer>` (uh...) or make the running averages of `<BatchNormLayer>` completely optional and just add a `<layernorm()>` convenience function that instantiates a `<BatchNormLayer>`. 

  I add a `<LayerNormLayer>`  which is very similar to `<BatchNormLayer>` and `<layer_norm>` as convenience function which is very similar to `<batch_norm>`. In the implementation, I just change the default `<axis>` to `<axes = tuple(range(0, len(self.input_shape)))>` and keep the `<batch_norm_use_averages=False, batch_norm_compute_averages=False>` part. 


- For the recurrent model, thanks for providing the `<LNLSTM>` model, which duplicates lots code from original LSTM. But I also have no good idea how to conduct the LN before gate activation without changing from internal code. Similarity, I have an implementation of `<LNGRU>' which is relatively simple than `<LNLSTM>`, but I don't think the LN of hidden_update activation is necessary. 

- For the usage, I use the mnist example (https://github.com/icemansina/LayerNorm/blob/master/test_mnist_for_ln.ipynb) for testing , please put the `<LayerNormLayer>`, `<layer_norm>`  and `<LNGRU>` into `<normalization>` and  `<recurrent>` respectively.   

-  If I am using cascaded CNN and RNN, would it influence the performance by using BN for CNN layer and LN for RNN layer according to your experience? Or I should use LN for both CNN and RNN layers. It seems very time-consuming. I am not sure which one is better. 

I know the recent code is very ugly and straightforward, please give me some suggestions. thanks a lot.

**Comment by erfannoury** [comment]
It is mentioned in the section 6.7 of the Layer Normalization paper that this normalization technique won't work as well as the Batch Normalization for Convolutional Layers. So I think you should stick with BN for convolutional layers and LN for recurrent layers.

Regarding the huge code duplication for implementing the LNLSTM, I couldn't find another way to not duplicate the functionality. LN internally changes the workings of the LSTM layer, so I think this was inevitable. However, it's definitely not planned for being added to the Lasagne.

**Comment by tkf** [comment]
Since I cannot find a complete implementation of @f0k's idea, here is my try. Unlike @icemansina's implementation, my version do not normalize over batch axis (0th axis) by default (`axes='auto'`). I believe this is the intention of Ba, Kiros & Hinton (2016). But I'm new to deep nets so I could be missing something.:

```python
import lasagne


class LayerNormLayer(lasagne.layers.BatchNormLayer):

    def __init__(self, incoming, axes='auto', **kwargs):
        if axes != 'auto':
            kwargs['axes'] = axes

        super(LayerNormLayer, self).__init__(
            incoming,
            beta=None, gamma=None,
            **kwargs)

        if axes == 'auto':
            self.axes = tuple(range(1, len(self.input_shape)))

    def get_output_for(self, input,
                       batch_norm_use_averages=False,
                       batch_norm_update_averages=False,
                       **kwargs):
        return super(LayerNormLayer, self).get_output_for(
            input,
            batch_norm_use_averages=batch_norm_use_averages,
            batch_norm_update_averages=batch_norm_update_averages,
            **kwargs)


def layer_normalized_dense_layer(incoming, num_units,
                                 nonlinearity=lasagne.nonlinearities.rectify,
                                 W=lasagne.init.Normal(std=1),
                                 b=lasagne.init.Constant(0.),
                                 **kwargs):
    assert num_units > 1
    layer = lasagne.layers.DenseLayer(
        incoming, num_units, W=W, b=None,
        nonlinearity=lasagne.nonlinearities.linear,
        **kwargs)
    layer = LayerNormLayer(layer)
    layer = lasagne.layers.ScaleLayer(layer)
    layer = lasagne.layers.BiasLayer(layer, b=b)
    return lasagne.layers.NonlinearityLayer(layer, nonlinearity=nonlinearity)
```

The above code works only for non-recurrent layers so I also have another very "hacky" implementation which relies on the (possibly) implementation detail of Lasagne that `nonlinearity` property can be array-to-array function rather than a scalar function:

```python
import lasagne
import theano


class LayerNormalization(object):

    def __init__(self, num_units,
                 nonlinearity=lasagne.nonlinearities.rectify,
                 b=lasagne.init.Constant(0.),
                 g=lasagne.init.Constant(1.),
                 eps=1e-5):
        self.num_units = num_units
        self.b = theano.shared(b.sample(num_units), name='layer_norm.b')
        self.g = theano.shared(g.sample(num_units), name='layer_norm.g')
        self.eps = eps
        self.nonlinearity = nonlinearity

    def normalizing_nonlinearity(self, x):
        mean = x.mean(-1, keepdims=True)
        sigma = theano.tensor.sqrt(x.var(-1, keepdims=True) + self.eps)
        b = self.b.reshape((1,) * (x.ndim - 1) + (-1,))
        g = self.g.reshape((1,) * (x.ndim - 1) + (-1,))
        return self.nonlinearity(g * (x - mean) / sigma + b)

    def register_to(self, layer):
        layer.add_param(self.b, (self.num_units,))
        layer.add_param(self.g, (self.num_units,))


class RecurrentNormalizingLayer(lasagne.layers.RecurrentLayer):

    def __init__(self, incoming, num_units,
                 W_hid_to_hid=lasagne.init.Normal(std=1),
                 b=lasagne.init.Uniform(0.1),
                 g=lasagne.init.Constant(1.),
                 hid_init=lasagne.init.Uniform(0.1),
                 nonlinearity=lasagne.nonlinearities.rectify,
                 eps=0.05,
                 **kwargs):
        self.layer_normalization = LayerNormalization(
            num_units, nonlinearity=nonlinearity, b=b, g=g, eps=eps)
        super(RecurrentNormalizingLayer, self).__init__(
            incoming, num_units,
            W_hid_to_hid=W_hid_to_hid,
            b=None,
            hid_init=hid_init,
            nonlinearity=self.layer_normalization.normalizing_nonlinearity,
            **kwargs)
        self.layer_normalization.register_to(self.hidden_to_hidden)
```

I guess the same idea can be used for LSTM and GRU.

Complete code with docstrings and tests can be found here:
https://gist.github.com/tkf/d74dc4efbaf6dc541838418741b05a2b
